sklearn/decomposition/pca.py

```diff
@@ -251,50 +327,75 @@ def fit_transform(self, X, y=None):
         return U
 
     def _fit(self, X):
-        """Fit the model on X
+        """Dispatch to the right submethod depending on the choice of solver.
 
-        Parameters
-        ----------
-        X: array-like, shape (n_samples, n_features)
-            Training vector, where n_samples in the number of samples and
-            n_features is the number of features.
+        The svd_solver='auto' case is expanded depending on the value of
+        n_components.
+        """
+        X = check_array(X, dtype=[np.float32, np.float64], ensure_2d=True,
+                        copy=self.copy)
 
-        Returns
-        -------
-        U, s, V : ndarrays
-            The SVD of the input data, copied and centered when
-            requested.
+        # Handle n_components==None
+        if self.n_components is None:
+            n_components = X.shape[1]
+        else:
+            n_components = self.n_components
+
+        # Handle svd_solver
+        svd_solver = self.svd_solver
+        if svd_solver == 'auto':
+            if n_components >= 1 and n_components < .8 * min(X.shape):
+                svd_solver = 'randomized'
+            else:
+                svd_solver = 'full'
+
+        # necessary to improve precision with few components
+        if svd_solver == 'randomized':
+            if n_components < .1 * min(X.shape):
+                self.iterated_power = max(self.iterated_power, 7)
```

Comment:
We should never mutate the constructor params in `fit` (only in `set_params`). Instead do:

```python
iterated_power = self.iterated_power
if svd_solver == 'randomized':
    if n_components < .1 * min(X.shape) and self.iterated_power > 7:
        iterated_power = 7
```

to avoid having to pass the new local variable `iterated_power` as an extra argument to `_fit_truncated` I think this block should be moved at least into the `_fit_truncated` method or maybe even into `randomized_range_finder`.

Also I think we should raise a `UserWarning` when we set `iterated_power = 7` based on this safety heuristic instead of doing it silently.